### ISSUE TITLE
"Directly injecting a ConfigProperty into a javax.ws.rs.ext.Provider may lead to unexpected results" warning for OidcClientRequestFilter in Quarkus 2.12

### ISSUE THREAD

**Issue by famod** [quarkusio/quarkus#27487]
### Describe the bug

Starting my app (in dev mode) logs the following message three times:
> 2022-08-24 21:07:54,704 WARN  [io.qua.arc.dep.ConfigBuildStep] (build-37) Directly injecting a org.eclipse.microprofile.config.inject.ConfigProperty into a javax.ws.rs.ext.Provider may lead to unexpected results. To ensure proper results, please change the type of the field to javax.enterprise.inject.Instance<java.util.Optional>. Offending field is 'clientName' of class 'io.quarkus.oidc.client.filter.OidcClientRequestFilter'

This wasn't there in 2.11.

Btw, we have two subclasses of that filter, e.g.:
```java
/**
 * Filter that configures using {@code foobar} {@link NamedOidcClient} for accessing a rest api.
 *
 * @see <a href="https://github.com/quarkusio/quarkus/issues/16397">Quarkus: Enhance OidcClientFilter to select named OIDC clients</a>
 */
public class FoobarOidcClientRequestFilter extends OidcClientRequestFilter {

    @Override
    protected Optional<String> clientId() {
        return Optional.of("foobar"); // has to match configured oidc-client id
    }
}
```

### Expected behavior

No such warning

### Actual behavior

Warning pointing to Quarkus code (not even my own one)

### How to Reproduce?

n/a for now

### Quarkus version or git rev

2.12.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.6

**Comment by sberyozkin** [comment]
Hmm, so right now it is
```
@Inject
@ConfigProperty(name = "quarkus.oidc-client-filter.client-name")
Optional<String> clientName;
```

not sure why it needs to be set as 

```
@Inject
@ConfigProperty(name = "quarkus.oidc-client-filter.client-name")
Instance<Optional<String>> clientName;
```

Looks like double optional...

**Comment by famod** [comment]
If I had to guess I'd say it's related to #26996.
/cc @radcortez

**Comment by geoand** [comment]
Although I agree that `Instance<Optional<String>>` looks pretty bad, I don't think however we can avoid it. I'll leave it to @radcortez however to confirm

**Comment by famod** [comment]
Mentioned classes from CI log (JVM Tests -JDK 11):
```
'clientName' of class 'io.quarkus.oidc.client.filter.OidcClientRequestFilter'
'config' of class 'io.quarkus.resteasy.test.config.ProviderConfigInjectionWarningsTest$FooProvider'
'configProperty' of class 'io.quarkus.resteasy.test.config.ProviderConfigInjectionWarningsTest$FooProvider'
'configProperty' of class 'io.quarkus.undertow.test.config.WebFilterConfigInjectionWarningsTest$ConfigFilter'
'exceptionConfig' of class 'io.quarkus.it.smallrye.config.ConfigurableExceptionMapper'
'mapping' of class 'io.quarkus.resteasy.test.config.ConfigMappingWithProviderTest$MappingFilter'
'message' of class 'io.quarkus.it.smallrye.config.ConfigurableExceptionMapper'
'myConfigMapping' of class 'io.quarkus.resteasy.test.config.ProviderConfigInjectionWarningsTest$FooProvider'
'smallRyeConfig' of class 'io.quarkus.resteasy.test.config.ProviderConfigInjectionWarningsTest$FooProvider'
```

**Comment by sberyozkin** [comment]
I wonder why `https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/OidcClientRequestReactiveFilter.java#L29` is not mentioned.

The other thing is, #26996 is related to native image - so having this warning in JVM mode is confusing as it gives no reasons - may be it can be reworded ?

I'm fine with adding an `Instant` - but would like Roberto to confirm it is required for Optional too.

**Comment by radcortez** [comment]
> I wonder why `https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/OidcClientRequestReactiveFilter.java#L29` is not mentioned.

Because the Reactive Filters are just regular beans and are not initialized at static init.

> The other thing is, #26996 is related to native image - so having this warning in JVM mode is confusing as it gives no reasons - may be it can be reworded ?

Well, if you build to native, you are stuck with the resolved config during the build for the filters in the native image, but it is not the only downside. Because provider filters are initialized at static init, they can only lookup configuration which is static init safe. For instance if you try to set `quarkus.oidc-client-filter.client-name` in a Vault Config Source, it won't work, because Vault is only added as a source for runtime, so this is also relevant for VM mode.

> I'm fine with adding an `Instance` - but would like Roberto to confirm it is required for Optional too.

If the property is `Optional` the `Instance` is still required. What `Instance` does is delay the call to the producer retrieving the configuration (usually, when the runtime instance is available). If there is no `Optional` then `Config` will treat the property as required and fail the lookup if the value is missing.

An alternative could be to provide the configuration in a side bean that is not initialized at static init.

Regardless, I think the warning, in this case, is correct, and the issue could occur, even if no one has complained about it yet.

I did detect another issue with the validation of `Instance<Optional>` injection points, but it is not related with the change in https://github.com/quarkusio/quarkus/pull/26996. I'll send a fix for it.

**Comment by radcortez** [comment]
Another alternative is to inject our root config object directly. Since it is build time and runtime fixed it is okay.

**Comment by radcortez** [comment]
> @radcortez is this issue really fixed? My understanding is that we needed to adjust our filters?

The `OidcClientRequestFilter` was the only case I could find in the production code. Everything else is testing code.

Since the `OidcClientRequestFilter` configuration is build time, it is okay. Users are not able to change the configuration un runtime. I don't think we can do anything else. RESTEasy Classic Factories (including the filters) are being created in Static  Init, and changing that may be problematic. I agree it should be more like Reactive.

**Comment by gsmet** [comment]
@radcortez but is the warning fixed?

**Comment by radcortez** [comment]
Yes, because we now use inject the ConfigRoot class directly, and we don't validate that:
https://github.com/radcortez/quarkus/blob/99e0134cf4cb9091a287d3cfd55508a181e73806/extensions/oidc-client-filter/runtime/src/main/java/io/quarkus/oidc/client/filter/OidcClientRequestFilter.java#L30

**Comment by gsmet** [comment]
OK, perfect then. Sorry for the noise :).

**Comment by radcortez** [comment]
No worries :)

**Comment by sberyozkin** [comment]
Thanks @radcortez

**Comment by famod** [comment]
Fix confirmed in 2.12.1, thanks!
